ranger-core/src/main/java/io/appform/ranger/core/finder/serviceregistry/ServiceRegistryUpdater.java

```diff
@@ -125,21 +130,42 @@ private Void queryExecutor() {
 
     private void updateRegistry() throws InterruptedException {
         log.debug("Checking for updates on data source for service: {}",
-                     serviceRegistry.getService().getServiceName());
-        if(!nodeDataSource.isActive()) {
-            log.warn("Node data source seems to be down. Keeping old list for {}",
-                        serviceRegistry.getService().getServiceName());
-            return;
-        }
-        val nodeList = nodeDataSource.refresh(deserializer).orElse(null);
-        if (null != nodeList) {
-            log.debug("Updating nodeList of size: {} for [{}]", nodeList.size(),
-                         serviceRegistry.getService().getServiceName());
-            serviceRegistry.updateNodes(nodeList);
+                  serviceRegistry.getService().getServiceName());
+        var callFailed = false;
+        if (nodeDataSource.isActive()) { //Source should implement circuit breaker to fail fast and reopen after some
+            // time
+            try {
+                val nodeList = nodeDataSource.refresh(deserializer).orElse(null);
+                if (null != nodeList) {
+                    log.debug("Updating nodeList of size: {} for [{}]", nodeList.size(),
+                              serviceRegistry.getService().getServiceName());
+                    val livenessCheckMaxAge = nodeDataSource.healthcheckZombieCheckThresholdTime(serviceRegistry.getService());
+                    //Remove all stale nodes before updating. This is done centrally to ensure some data sources
+                    //don't skip this check. Some control is still provided so that they can overload.
+                    serviceRegistry.updateNodes(FinderUtils.filterValidNodes(serviceRegistry.getService(), nodeList, livenessCheckMaxAge));
+                }
+                else {
+                    log.warn("Empty list returned from node data source. We are in a weird state. Keeping old list for {}",
```

Comment:
The nodeDataSource.refresh on callFailure - returns null, where the nodeList is retained as is, already. So that shouldn't be a problem. 

I don't know what you mean by nodeDataStore status? @ToOnlyGaurav 

TLDR: Old list is returned and maintained if the backend is not responding. And empty list is only returned by the nodeDataSource, when backend truly returns an empty list or the nodes are stale or unhealthy.